codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -69,409 +71,1060 @@
+        // channelReadComplete(...)
+        DefaultHttp2StreamChannel next;
+
+        DefaultHttp2StreamChannel(Http2FrameStream stream) {
+            this.stream = stream;
+            ((Http2MultiplexCodecStream) stream).channel = this;
+            pipeline = new DefaultChannelPipeline(this) {
+                @Override
+                protected void incrementPendingOutboundBytes(long size) {
+                    // Do thing for now
                 }
-            }
+
+                @Override
+                protected void decrementPendingOutboundBytes(long size) {
+                    // Do thing for now
+                }
+            };
+            closePromise = pipeline.newPromise();
+            channelId = new Http2StreamChannelId(parent().id(), ++idCount);
         }
-    }
 
-    @SuppressWarnings("unchecked")
-    private static void initAttrs(Channel channel, Map<AttributeKey<?>, Object> attrs) {
-        if (attrs != null) {
-            for (Entry<AttributeKey<?>, Object> e: attrs.entrySet()) {
-                channel.attr((AttributeKey<Object>) e.getKey()).set(e.getValue());
+        @Override
+        public Http2FrameStream stream() {
+            return stream;
+        }
+
+        void streamClosed() {
+            streamClosedWithoutError = true;
+            if (readInProgress) {
+                // Just call closeForcibly() as this will take care of fireChannelInactive().
+                unsafe().closeForcibly();
+            } else {
+                closePending = true;
             }
         }
-    }
 
-    final class Http2StreamChannel extends AbstractHttp2StreamChannel implements ChannelFutureListener {
-        boolean onStreamClosedFired;
+        @Override
+        public ChannelMetadata metadata() {
+            return METADATA;
+        }
 
-        /**
-         * {@code true} if stream is in {@link Http2MultiplexCodec#channelsToFireChildReadComplete}.
-         */
-        boolean inStreamsToFireChildReadComplete;
+        @Override
+        public ChannelConfig config() {
+            return config;
+        }
 
-        Http2StreamChannel(Channel parentChannel) {
-            super(parentChannel);
+        @Override
+        public boolean isOpen() {
+            return !closePromise.isDone();
         }
 
         @Override
-        protected void doClose() throws Exception {
-            if (!onStreamClosedFired && isStreamIdValid(streamId())) {
-                Http2StreamFrame resetFrame = new DefaultHttp2ResetFrame(Http2Error.CANCEL).streamId(streamId());
-                writeFromStreamChannel(resetFrame, true);
-            }
-            super.doClose();
+        public boolean isActive() {
+            return isOpen();
         }
 
         @Override
-        protected void doWrite(Object msg) {
+        public boolean isWritable() {
+            // TODO: Should we also take the parent channel into account ?
+            return isStreamIdValid(stream.id()) && writable;
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)